File: ios/RecipeApp/Services/RecipeService.swift
import Foundation

class RecipeService {
    static let shared = RecipeService()
    private let client = APIClient.shared

    private init() {}

    // MARK: - Recipe CRUD

    func createRecipe(_ request: CreateRecipeRequest) async throws -> Recipe {
        let response: RecipeResponse = try await client.request(
            endpoint: "/recipes",
            method: "POST",
            body: request
        )
        return response.recipe
    }

    func getRecipes(limit: Int = 50, offset: Int = 0, sortBy: String = "created_at", order: String = "DESC") async throws -> [Recipe] {
        let response: RecipesResponse = try await client.request(
            endpoint: "/recipes?limit=\(limit)&offset=\(offset)&sortBy=\(sortBy)&order=\(order)",
            method: "GET"
        )
        return response.recipes
    }

    func getRecipe(id: UUID) async throws -> Recipe {
        let response: RecipeResponse = try await client.request(
            endpoint: "/recipes/\(id.uuidString)",
            method: "GET"
        )
        return response.recipe
    }

    func updateRecipe(id: UUID, request: CreateRecipeRequest) async throws -> Recipe {
        let response: RecipeResponse = try await client.request(
            endpoint: "/recipes/\(id.uuidString)",
            method: "PUT",
            body: request
        )
        return response.recipe
    }

    func deleteRecipe(id: UUID) async throws {
        let _: EmptyResponse = try await client.request(
            endpoint: "/recipes/\(id.uuidString)",
            method: "DELETE"
        )
    }

    func searchRecipes(query: String) async throws -> [Recipe] {
        let encodedQuery = query.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? query
        let response: RecipesResponse = try await client.request(
            endpoint: "/recipes/search?q=\(encodedQuery)",
            method: "GET"
        )
        return response.recipes
    }

    func toggleFavorite(id: UUID) async throws -> Recipe {
        let response: RecipeResponse = try await client.request(
            endpoint: "/recipes/\(id.uuidString)/favorite",
            method: "POST"
        )
        return response.recipe
    }

    // MARK: - Images

    func uploadImages(recipeId: UUID, images: [Data]) async throws -> RecipeResponse {
        return try await client.uploadImages(
            endpoint: "/recipes/\(recipeId.uuidString)/images",
            images: images
        )
    }

    func deleteImage(recipeId: UUID, imageId: UUID) async throws {
        let _: EmptyResponse = try await client.request(
            endpoint: "/recipes/\(recipeId.uuidString)/images/\(imageId.uuidString)",
            method: "DELETE"
        )
    }

    func setPrimaryImage(recipeId: UUID, imageId: UUID) async throws {
        let _: EmptyResponse = try await client.request(
            endpoint: "/recipes/\(recipeId.uuidString)/images/\(imageId.uuidString)/primary",
            method: "PUT"
        )
    }

    // MARK: - Daily Recipe

    func getDailyRecipe() async throws -> Recipe {
        let response: DailyRecipeResponse = try await client.request(
            endpoint: "/recipe-of-the-day",
            method: "GET",
            requiresAuth: false
        )
        return response.recipe
    }

    func getDailyRecipeHistory() async throws -> [DailyRecipeHistoryItem] {
        let response: DailyRecipeHistory = try await client.request(
            endpoint: "/recipe-of-the-day/history",
            method: "GET",
            requiresAuth: false
        )
        return response.history
    }
}

File: web/src/services/recipe.service.ts
import { api } from './api';
import { Recipe, CreateRecipeRequest } from '@/types';

interface RecipesResponse {
  recipes: Recipe[];
  count: number;
}

interface RecipeResponse {
  recipe: Recipe;
  message?: string;
}

export const recipeService = {
  async getRecipes(params?: {
    limit?: number;
    offset?: number;
    sortBy?: string;
    order?: string;
  }): Promise<Recipe[]> {
    const queryParams = new URLSearchParams(params as any).toString();
    const response = await api.get<RecipesResponse>(
      `/recipes${queryParams ? `?${queryParams}` : ''}`
    );
    return response.recipes;
  },

  async getRecipe(id: string): Promise<Recipe> {
    const response = await api.get<RecipeResponse>(`/recipes/${id}`);
    return response.recipe;
  },

  async createRecipe(data: CreateRecipeRequest): Promise<Recipe> {
    const response = await api.post<RecipeResponse>('/recipes', data);
    return response.recipe;
  },

  async updateRecipe(id: string, data: Partial<CreateRecipeRequest>): Promise<Recipe> {
    const response = await api.put<RecipeResponse>(`/recipes/${id}`, data);
    return response.recipe;
  },

  async deleteRecipe(id: string): Promise<void> {
    await api.delete(`/recipes/${id}`);
  },

  async searchRecipes(query: string): Promise<Recipe[]> {
    const response = await api.get<RecipesResponse>(`/recipes/search?q=${encodeURIComponent(query)}`);
    return response.recipes;
  },

  async toggleFavorite(id: string): Promise<Recipe> {
    const response = await api.post<RecipeResponse>(`/recipes/${id}/favorite`);
    return response.recipe;
  },

  async uploadImages(recipeId: string, images: File[]): Promise<void> {
    const formData = new FormData();
    images.forEach((image) => {
      formData.append('images', image);
    });
    await api.uploadFormData(`/recipes/${recipeId}/images`, formData);
  },

  async getDailyRecipe(): Promise<Recipe> {
    const response = await api.get<{ recipe: Recipe }>('/recipe-of-the-day');
    return response.recipe;
  },
};
